### PR TITLE
Move Copter's CHn_OPT to be RCn_OPTION

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by peterbarker** [comment]
This was mentioned in #6811

**Comment by OXINARF** [comment]
I'm against adding the RCn_OPTION in RC_Channel until we have a way to document values for different vehicles (and a way to say that a parameter only applies to certain vehicles is also needed). The SERVOn_FUNCTION is already messy with all the values, making the same for the options would be even worse.

**Comment by peterbarker** [comment]
@OXINARF But if we broke out different values for different vehicles that would be *inconsistent* :-)

We actually threw this around the other night without coming to a concrete conclusion.

What if we were to have `@Values[Copter]` and `@Values[Plane]`?

**Comment by OXINARF** [comment]
> But if we broke out different values for different vehicles that would be inconsistent :-)

It's supposed to be inconsistent, given that vehicles don't all support the same stuff 🙂

> What if we were to have @Values[Copter] and @Values[Plane]?

Yes, that's kind of what I was thinking. If a @Values exists and @Values[Copter] doesn't, then Copter would use the @Values one - that would allow only one vehicle to be different for example.
This should also exist for @Range and @Description.

Also, (at least) Plane doesn't use OPT, so this parameter shouldn't be displayed there (which should be easy now with the parameter options), but it should also not be present in documentation or people will be confused - hence the need to have a way to say to what vehicles it applies (understanding heli frame would be worth too - would fix the duplicated parameters we have in documentation now - but that's the extra mile).

**Comment by amilcarlucas** [comment]
> understanding heli frame would be worth too - would fix the duplicated parameters we have in documentation now - but that's the extra mile

That would be the icing on the cake :)

**Comment by amilcarlucas** [comment]
If the mode becomes a ch option than IMHO this PR fixes https://github.com/ArduPilot/ardupilot/issues/1974 and https://github.com/ArduPilot/ardupilot/issues/3819 :)

**Comment by peterbarker** [comment]
@OXINARF I've pushed up changes so that:
 - `RCn_OPTION` won't appear anywhere but Copter for the time being
 - we only ever generate parameters for a single vehicle at a time
 - parameter values can be filtered by {tag}

**Comment by peterbarker** [comment]
@OXINARF Nudge

**Comment by amilcarlucas** [comment]
I think we now have a way to document values for different vehicles. @peterbarker Can you update this PR to make use of it ?

**Comment by amilcarlucas** [comment]
On a side note: Shouldn't FS_THROTTLE should be renamed to RC_FS ?

**Comment by peterbarker** [comment]
On Sat, 23 Dec 2017, Amilcar Lucas wrote:

> I think we now have a way to document values for different vehicles.
> @peterbarker Can you update this PR to make use of it ?

It already does :-)

https://github.com/ArduPilot/ardupilot/pull/6819/files#diff-dec3f64931c56a074db3501032fe7e6aR76

If you look through the history of the PR it used to contain the changes 
to the parameter generation.  That was split into a separate PR and this 
was rebased on top of that branch....

**Comment by peterbarker** [comment]
On Sat, 23 Dec 2017, Amilcar Lucas wrote:

> On a side note: Shouldn't FS_THROTTLE should be renamed to RC_FS ?

That's an interesting idea.  Unrelated to this ticket, 'though, and 
currently each vehicle handles their own failsafes, pretty much.

Peter
y

**Comment by lvale** [comment]
@peterbarker And what about having all the RC channels being able to have 3 states (1000/1500/2000) with a range of 100 around each value ? This would expand considerably the usability of Options and simplify a lot the latest radio programmings

**Comment by peterbarker** [comment]
On Sat, 23 Dec 2017, Luis Vale Gonçalves wrote:

> @peterbarker And what about having all the RC channels being able to have 3
> states (1000/1500/2000) with a range of 100 around each value ? This would
> expand considerably the usability of Options and simplify a lot the latest
> radio programmings

Why just three?

This is complicated by the fact that these "option" switches cope with 
three-positions already, which is used by the simple/supersimple modes, 
amongst other things.

That means we can't just throw in more options like RC7_OPTION2 (which 
would allow you to specify an action for between, say 1500-1600 PWM), as 
then we'd have a conflcit in configuration if RC7_OPTION specified 
simple/supersimple switch and RC7_OPTION was anything other than 
"nothing".  We could add sanity prearm checks to catch these, but they'd 
be arbitrarily complicated.

Is this an actual problem?  Is it something that needs to be even brought 
up in relation to this PR?

**Comment by lvale** [comment]
@peterbarker 
well, three because most of the RC radios don't have more than 3 position switches, and/or 2 position switches. As an example the Taranis can be modified with a 6 position switch but that is an exception and requires some DIY...

Best of my knowledge, the Simple/SuperSimple is the only option that can be used with a 3 pos switch.

Let's take a regular RC radio like the FrSky Taranis that can handle via SBus 16 channels, as an example.

ArduPilot reserves 5 channels (AIL,EL,TH,RD + mode)

No problem here, we still have 11(?) free channels

Add a Gimbal, camera and a Landing Gear

3 to 5 channels can easily go to a gimbal (Roll/Pitch/Yaw/Picture/Video), and one extra to the LGR

Add some channels to Flight modifiers, like AutoTune/Simple/Break/Kill Switch/Land/RTL/SmartRTL/LostCopter/ArmDisarm/Throw/Auto/SaveWP that's 12 more...

The issue is, that we're adding functionality, but we don't expose those functions easily to the user.

On highly programmable radios (like the Taranis) I can use a crazy mix of combos assigned to the switches so I can get by with my settings easily, but although we have a ever increasing number of available options we tend to force the users to only use a limited number of them.

File: ArduCopter/Parameters.h [hunk 1]
```diff
@@ -210,10 +210,10 @@ class Parameters {
         k_param_serial2_baud_old, // deprecated
         k_param_serial2_protocol, // deprecated
         k_param_serial_manager,
-        k_param_ch9_option,
-        k_param_ch10_option,
-        k_param_ch11_option,
-        k_param_ch12_option,
+        k_param_ch9_option_old,
+        k_param_ch10_option_old,
+        k_param_ch11_option_old,
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduCopter/Parameters.h]: Yes - and I'll add a commit to do the same for all of the other `_old` parameter variables!

[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduCopter/Parameters.h]: Whoops.  No.  Required for conversion code.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 2]
```diff
@@ -70,6 +70,13 @@ const AP_Param::GroupInfo RC_Channel::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("DZ",   5, RC_Channel, dead_zone, 0),
 
+    // @Param: OPTION
+    // @DisplayName: RC input option
+    // @Description: Function assigned to this RC channel
+    // @Values{ArduCopter}: 0:Do Nothing, 2:Flip, 3:Simple Mode, 4:RTL, 5:Save Trim, 7:Save WP, 9:Camera Trigger, 10:RangeFinder, 11:Fence, 13:Super Simple Mode, 14:Acro Trainer, 15:Sprayer, 16:Auto, 17:AutoTune, 18:Land, 19:Gripper, 21:Parachute Enable, 22:Parachute Release, 23:Parachute 3pos, 24:Auto Mission Reset, 25:AttCon Feed Forward, 26:AttCon Accel Limits, 27:Retract Mount, 28:Relay On/Off, 34:Relay2 On/Off, 35:Relay3 On/Off, 36:Relay4 On/Off, 29:Landing Gear, 30:Lost Copter Sound, 31:Motor Emergency Stop, 32:Motor Interlock, 33:Brake, 37:Throw, 38:ADSB-Avoidance, 39:PrecLoiter, 40:Object Avoidance, 41:ArmDisarm, 42:SmartRTL, 43:InvertedFlight, 44:Winch Enable, 45:WinchControl
```
[the remark this answers is not in the document]
Comment (peterbarker) [reply libraries/RC_Channel/RC_Channel.cpp]: Once this is in *then* I can convert those!

File: ArduCopter/AP_Arming.cpp [hunk 3]
```diff
@@ -185,7 +186,7 @@ bool AP_Arming_Copter::parameter_checks(bool display_failure)
     if ((checks_to_perform == ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_PARAMETERS)) {
 
         // ensure ch7 and ch8 have different functions
```
Comment (khancyr) [inline ArduCopter/AP_Arming.cpp]: check that channels have different functions

File: ArduCopter/switches.cpp [hunk 4]
```diff
@@ -3,15 +3,20 @@
...
-    uint8_t CH8_flag    : 2; // 4, 5    // ch8 aux switch : 0 is low or false, 1 is center or true, 2 is high
-    uint8_t CH9_flag    : 2; // 6, 7    // ch9 aux switch : 0 is low or false, 1 is center or true, 2 is high
-    uint8_t CH10_flag   : 2; // 8, 9    // ch10 aux switch : 0 is low or false, 1 is center or true, 2 is high
-    uint8_t CH11_flag   : 2; // 10,11   // ch11 aux switch : 0 is low or false, 1 is center or true, 2 is high
-    uint8_t CH12_flag   : 2; // 12,13   // ch12 aux switch : 0 is low or false, 1 is center or true, 2 is high
-} aux_con;
+// 0 is low or false, 1 is center or true, 2 is high
+// pairs of bits in old_switch_positions give the old switch position for an RC input.
+uint32_t old_switch_positions;
+uint8_t old_switch_position(const uint8_t chan);
```
Comment (khancyr) [inline ArduCopter/switches.cpp]: those should get back on copter.h

Comment (khancyr) [reply ArduCopter/switches.cpp]: or stay here until we merge the AP_Switch PR

File: ArduCopter/switches.cpp [hunk 5]
```diff
@@ -3,15 +3,20 @@
...
+uint8_t old_switch_position(const uint8_t chan)
+{
+    return ((old_switch_positions >> (chan*2)) & 0x3);
+}
+void set_old_switch_position(const uint8_t chan, const uint8_t value);
+void set_old_switch_position(const uint8_t chan, const uint8_t value)
+{
+    old_switch_positions &= ~(0x3<<(chan*2));
```
Comment (khancyr) [inline ArduCopter/switches.cpp]: miss spaces around << and *

**Comment by OXINARF** [comment]
@peterbarker This still has the generation of the parameter for vehicles where it doesn't apply right? It won't have values, but it will show up in documentation - unless I missed something.

File: ArduCopter/AP_Arming.cpp (hunk 3 above)
Comment (peterbarker) [reply ArduCopter/AP_Arming.cpp]: Fixed

**Comment by peterbarker** [comment]
On Tue, 20 Mar 2018, Pierre Kancir wrote:

> +uint32_t old_switch_positions;
> +uint8_t old_switch_position(const uint8_t chan);
> 
> or stay here until we merge the AP_Switch PR

Yes!

> +{
> +    old_switch_positions &= ~(0x3<<(chan*2));
> 
> miss spaces around << and *

Adding.

**Comment by peterbarker** [comment]
@OXINARF They do get generated, and they are empty.

**Comment by robustini** [comment]
Epic improvements @peterbarker, congrats!

**Comment by OXINARF** [comment]
@peterbarker OK, so you are going to leave it that way? I think that's going to be confusing for the users.

**Comment by peterbarker** [comment]
On Wed, 21 Mar 2018, Francisco Ferreira wrote:

> @peterbarker OK, so you are going to leave it that way? I think that's going
> to be confusing for the users.

I could add another piece of markup to the code, I guess.

I think I'm going to have to, really.

File: ArduCopter/Parameters.cpp [hunk 6]
```diff
@@ -1057,6 +1015,13 @@ const AP_Param::ConversionInfo conversion_table[] = {
...
     { Parameters::k_param_serial2_baud,       0,      AP_PARAM_INT16, "SERIAL2_BAUD" },
     { Parameters::k_param_arming_check_old,   0,      AP_PARAM_INT8,  "ARMING_CHECK" },
+
+    { Parameters::Parameters::k_param_ch7_option_old,   0,      AP_PARAM_INT8,  "RC7_OPTION" },
+    { Parameters::Parameters::k_param_ch8_option_old,   0,      AP_PARAM_INT8,  "RC8_OPTION" },
+    { Parameters::Parameters::k_param_ch9_option_old,   0,      AP_PARAM_INT8,  "RC9_OPTION" },
+    { Parameters::Parameters::k_param_ch10_option_old,   0,      AP_PARAM_INT8,  "RC10_OPTION" },
+    { Parameters::Parameters::k_param_ch11_option_old,   0,      AP_PARAM_INT8,  "RC11_OPTION" },
+    { Parameters::Parameters::k_param_ch12_option_old,   0,      AP_PARAM_INT8,  "RC12_OPTION" },
 };
```
Comment (lvale) [inline ArduCopter/Parameters.cpp]: still missing RC_13, RC_14, RC_15 and RC_16

:)

**Comment by peterbarker** [comment]
On Thu, 22 Mar 2018, Luis Vale Gonçalves wrote:

> still missing RC_13, RC_14, RC_15 and RC_16
> 
> :)

That's conversion code.  We can't convert parameters that don't exist...

**Comment by peterbarker** [comment]
On Thu, 22 Mar 2018, Peter Barker wrote:

> I could add another piece of markup to the code, I guess.
> I think I'm going to have to, really.

No extra markup required.  Just changed parameter generation script to not 
show anything if Values{thisvehiclename}: isn't present, 
Values{anyothervehiclename}: is, and there's no generic Values:

From an actual-parameters perspective these won't be shown to a GCS 
because of the AP_PARAM_FRAME_COPTER in the parameter definition.

**Comment by amilcarlucas** [comment]
Are all issues are addressed now ? Can this be merged ?

**Comment by peterbarker** [comment]
https://github.com/peterbarker/ardupilot/tree/chopt-to-rcinput-rover <- that branch contains a rough conversion of Rover to using `RCn_OPTION` in place of its single `AUX_CH` parameter.

Still to do: parameter conversion,  sanity checks.

**Comment by khancyr** [comment]
@peterbarker what is miss is the conversion for other vehicle right ?

**Comment by peterbarker** [comment]
On Thu, 12 Apr 2018, Pierre Kancir wrote:

> @peterbarker what is miss is the conversion for other vehicle right ?

I was going to do the other vehicles after this one.

Doing all at the same time simply means when I rebase this PR I have to 
solve four times the conflicts in the vehicle directories.  It also means 
getting approval from the other vehicle maintainers on a single PR.

**Comment by rmackay9** [comment]
Closing this one 'cuz it's included in https://github.com/ArduPilot/ardupilot/pull/8170

**Pull request closed without merging**